File: browser-extension/ibm-literature-capture/native-bridge/host.py
#!/usr/bin/env python3
"""iBM Lab Native Messaging 下载桥（开发回退版）。

只接受 upload 命令：读取浏览器已完成且位于批准下载目录内的文件，并上传到
本机 iBM Lab Agent 的一次性捕获接口。生产桌面安装包使用同协议的 Node.js Host。
"""

import json
import os
import re
import struct
import sys
import urllib.parse
import urllib.request

BRIDGE_NAME = "com.ibm.lab.capture"

# ── 捕获安全 spec（单点事实源）────────────────────────────────────────────
# 与 background.js、desktop/src-tauri/resources/bridge/host.js 共用同一份 JSON
# （browser-extension/…/capture-spec.json），避免安全边界各自维护而漂移。
# 找不到 spec 时直接失败（fail-closed），不回退到“宽容”的旧常量。
_CAPTURE_SPEC_FILE = os.path.join(
    os.path.dirname(os.path.dirname(os.path.abspath(__file__))),
    "capture-spec.json",
)


def _load_capture_spec():
    try:
        with open(_CAPTURE_SPEC_FILE, "r", encoding="utf-8") as handle:
            return json.load(handle)
    except Exception as exc:  # fail closed
        raise RuntimeError(
            "cannot load capture security spec %s (%s)" % (_CAPTURE_SPEC_FILE, exc)
        ) from exc


_spec = _load_capture_spec()
MAX_FILE_BYTES = int(_spec.get("maxFileBytes", 100 * 1024 * 1024))
CAPTURE_UPLOAD_PATH = str(_spec.get("captureUploadPath", "/api/lab-capture-upload"))
LOOPBACK_HOSTS = set(_spec.get("loopbackHosts", ["127.0.0.1", "localhost", "::1"]))
TASK_ID_RE = re.compile(str(_spec.get("taskIdPattern", r"^capture-[a-z0-9]+$")))


def configure_binary_stdio():
    if os.name == "nt":
        import msvcrt
        msvcrt.setmode(sys.stdin.fileno(), os.O_BINARY)
        msvcrt.setmode(sys.stdout.fileno(), os.O_BINARY)


def read_message():
    raw = sys.stdin.buffer.read(4)
    if len(raw) < 4:
        return None
    (length,) = struct.unpack("@I", raw)
    if length <= 0 or length > 1024 * 1024:
        return None
    return json.loads(sys.stdin.buffer.read(length).decode("utf-8"))


def write_message(value):
    data = json.dumps(value, ensure_ascii=False).encode("utf-8")
    sys.stdout.buffer.write(struct.pack("@I", len(data)))
    sys.stdout.buffer.write(data)
    sys.stdout.buffer.flush()


def _browser_download_dirs():
    found = []
    roots = [
        os.path.join(os.environ.get("LOCALAPPDATA", ""), "Google", "Chrome", "User Data"),
        os.path.join(os.environ.get("LOCALAPPDATA", ""), "Microsoft", "Edge", "User Data"),
    ]
    for root in roots:
        if not os.path.isdir(root):
            continue
        for name in os.listdir(root):
            preferences = os.path.join(root, name, "Preferences")
            if not os.path.isfile(preferences):
                continue
            try:
                with open(preferences, "r", encoding="utf-8") as handle:
                    directory = (json.load(handle).get("download", {}) or {}).get("default_directory")
                if directory:
                    found.append(os.path.expandvars(directory))
            except Exception:
                continue
    return found


def downloads_dirs():
    dirs = _browser_download_dirs()
    try:
        import winreg
        key_path = r"Software\Microsoft\Windows\CurrentVersion\Explorer\User Shell Folders"
        with winreg.OpenKey(winreg.HKEY_CURRENT_USER, key_path) as handle:
            value, _ = winreg.QueryValueEx(handle, "{374DE290-123F-4565-9164-39C4925E467B}")
        dirs.append(os.path.expandvars(value))
    except Exception:
        pass
    home = os.environ.get("USERPROFILE") or os.path.expanduser("~")
    dirs.extend([os.path.join(home, name) for name in ("Downloads", "下载", "Desktop", "桌面")])
    seen = set()
    result = []
    for directory in dirs:
        key = os.path.normcase(os.path.abspath(directory))
        if directory and os.path.isdir(directory) and key not in seen:
            seen.add(key)
            result.append(directory)
    return result


def _inside_directory(candidate, base):
    try:
        return os.path.commonpath([
            os.path.normcase(os.path.realpath(candidate)),
            os.path.normcase(os.path.realpath(base)),
        ]) == os.path.normcase(os.path.realpath(base))
    except ValueError:
        return False


def resolve_download_path(download_path):
    if not download_path:
        return None, "empty download path"
    normalized = os.path.normpath(download_path.replace("/", os.sep).replace("\\", os.sep))
    bases = downloads_dirs()
    candidates = [normalized] if os.path.isabs(normalized) else [os.path.join(base, normalized) for base in bases]
    for candidate in candidates:
        absolute = os.path.abspath(candidate)
        if any(_inside_directory(absolute, base) for base in bases) and os.path.isfile(absolute):
            return absolute, None
    return None, "file is missing or outside approved download directories: " + download_path


def validate_upload_url(value):
    parsed = urllib.parse.urlparse(str(value or ""))
    query = urllib.parse.parse_qs(parsed.query, keep_blank_values=True)
    if parsed.scheme != "http" or (parsed.hostname or "").lower() not in LOOPBACK_HOSTS:
        raise ValueError("uploadUrl is not a loopback iBM capture endpoint")
    if parsed.path != CAPTURE_UPLOAD_PATH or parsed.username or parsed.password or parsed.fragment:
        raise ValueError("uploadUrl contains unauthorized components")
    if set(query) != {"token"} or len(query["token"]) != 1 or not query["token"][0]:
        raise ValueError("uploadUrl must contain exactly one capture token")
    return value


def upload(upload_url, file_path, file_name):
    validate_upload_url(upload_url)
    size = os.path.getsize(file_path)
    if size < 1 or size > MAX_FILE_BYTES:
        raise ValueError("file size is outside the capture limit")
    with open(file_path, "rb") as handle:
        content = handle.read()
    request = urllib.request.Request(upload_url, data=content, method="PUT")
    request.add_header("Content-Type", "application/octet-stream")
    request.add_header("X-File-Name", urllib.parse.quote(file_name))
    request.add_header("X-Capture-Source", "native-bridge")
    request.add_header("Content-Length", str(len(content)))
    with urllib.request.urlopen(request, timeout=300) as response:
        body = response.read().decode("utf-8", "replace")
    try:
        return json.loads(body)
    except ValueError:
        return {"ok": False, "error": "server returned non-JSON response: " + body[:200]}


def main():
    configure_binary_stdio()
    message = read_message()
    if not message:
        write_message({"ok": False, "error": "empty message"})
        return
    if message.get("cmd") != "upload":
        write_message({"ok": False, "error": "unsupported command: download bridge accepts upload only"})
        return
    task_id = str(message.get("taskId", ""))
    upload_url = str(message.get("uploadUrl", ""))
    download_path = str(message.get("downloadPath", "") or message.get("relativePath", ""))
    file_name = str(message.get("fileName", "") or os.path.basename(download_path.replace("\\", "/")))
    if not TASK_ID_RE.fullmatch(task_id):
        write_message({"ok": False, "error": "invalid taskId"})
        return
    try:
        validate_upload_url(upload_url)
    except Exception as error:
        write_message({"ok": False, "taskId": task_id, "error": str(error)})
        return
    file_path, path_error = resolve_download_path(download_path)
    if path_error:
        write_message({"ok": False, "taskId": task_id, "error": path_error})
        return
    try:
        response = upload(upload_url, file_path, file_name)
        ok = bool(response.get("ok"))
        write_message({"ok": ok, "taskId": task_id, "response": response,
                       "error": None if ok else response.get("error", "upload rejected")})
    except Exception as error:
        write_message({"ok": False, "taskId": task_id, "error": str(error)})


if __name__ == "__main__":
    main()

File: desktop/src-tauri/resources/bridge/host.js
#!/usr/bin/env node
/**
 * iBM Lab Native Messaging 下载桥（Node.js，零依赖）。
 *
 * 只接受 upload 命令：读取 Edge/Chrome 已完成且位于批准下载目录内的文件，
 * 上传到本机 iBM Lab Agent 的一次性捕获接口。项目目录由服务端任务决定，扩展和
 * Native Host 都不能指定目标目录。不得加入通用下载、Office 保存或任意 URL 请求。
 */

"use strict";

const fs = require("node:fs");
const http = require("node:http");
const path = require("node:path");
const { execFileSync } = require("node:child_process");

const BRIDGE_NAME = "com.ibm.lab.capture";
const MAX_FILE_BYTES = 100 * 1024 * 1024;
const CAPTURE_UPLOAD_PATH = "/api/lab-capture-upload";
const LOOPBACK_HOSTS = new Set(["127.0.0.1", "localhost", "::1"]);
const TASK_ID_RE = /^capture-[a-z0-9]+$/;

function readFully(buffer) {
  let filled = 0;
  while (filled < buffer.length) {
    const got = fs.readSync(process.stdin.fd, buffer, filled, buffer.length - filled, null);
    if (got <= 0) return false;
    filled += got;
  }
  return true;
}

function readMessage() {
  const frame = Buffer.alloc(4);
  if (!readFully(frame)) return null;
  const payloadLength = frame.readUInt32LE(0);
  if (payloadLength <= 0 || payloadLength > 1024 * 1024) return null;
  const payload = Buffer.alloc(payloadLength);
  if (!readFully(payload)) return null;
  return JSON.parse(payload.toString("utf8"));
}

function writeMessage(value) {
  const data = Buffer.from(JSON.stringify(value), "utf8");
  const frame = Buffer.alloc(4);
  frame.writeUInt32LE(data.length, 0);
  process.stdout.write(frame);
  process.stdout.write(data);
}

function log(message) {
  process.stderr.write(`[${BRIDGE_NAME}] ${message}\n`);
}

function expandEnv(value) {
  return String(value).replace(/%([^%]+)%/g, (_, name) => process.env[name] ?? `%${name}%`);
}

function browserDownloadDirs() {
  const found = [];
  const roots = [
    path.join(process.env.LOCALAPPDATA || "", "Google", "Chrome", "User Data"),
    path.join(process.env.LOCALAPPDATA || "", "Microsoft", "Edge", "User Data")
  ];
  for (const root of roots) {
    let entries;
    try { entries = fs.readdirSync(root, { withFileTypes: true }); } catch { continue; }
    for (const entry of entries) {
      if (!entry.isDirectory()) continue;
      try {
        const preferences = JSON.parse(fs.readFileSync(path.join(root, entry.name, "Preferences"), "utf8"));
        const directory = preferences?.download?.default_directory;
        if (directory) found.push(expandEnv(directory));
      } catch { /* 其他 profile 或锁定文件忽略 */ }
    }
  }
  return found;
}

function shellDownloadsDir() {
  try {
    const raw = execFileSync(
      "reg",
      ["query", "HKCU\\Software\\Microsoft\\Windows\\CurrentVersion\\Explorer\\User Shell Folders", "/v", "{374DE290-123F-4565-9164-39C4925E467B}"],
      { encoding: "buffer", windowsHide: true, timeout: 5000 }
    );
    const match = new TextDecoder("gbk").decode(raw).match(/REG_(?:EXPAND_)?SZ\s+(.+)$/m);
    if (match) return expandEnv(match[1].trim());
  } catch { /* 注册表不可用时走常见目录 */ }
  return null;
}

function downloadsDirs() {
  const dirs = browserDownloadDirs();
  const shell = shellDownloadsDir();
  if (shell) dirs.push(shell);
  const home = process.env.USERPROFILE || process.env.HOME || "";
  dirs.push(path.join(home, "Downloads"), path.join(home, "下载"), path.join(home, "Desktop"), path.join(home, "桌面"));
  const seen = new Set();
  return dirs.filter((directory) => {
    if (!directory) return false;
    const key = path.resolve(directory).toLowerCase();
    if (seen.has(key)) return false;
    seen.add(key);
    try { return fs.statSync(directory).isDirectory(); } catch { return false; }
  });
}

function insideDirectory(candidate, base) {
  const candidatePath = path.resolve(candidate).toLowerCase();
  const basePath = path.resolve(base).toLowerCase();
  return candidatePath === basePath || candidatePath.startsWith(basePath.endsWith(path.sep) ? basePath : basePath + path.sep);
}

function resolveDownloadPath(downloadPath) {
  if (!downloadPath) return { path: null, error: "empty download path" };
  const normalized = path.normalize(String(downloadPath).replace(/[/\\]/g, path.sep));
  const bases = downloadsDirs();
  const candidates = path.isAbsolute(normalized) ? [normalized] : bases.map((base) => path.join(base, normalized));
  for (const candidate of candidates) {
    const absolute = path.resolve(candidate);
    if (!bases.some((base) => insideDirectory(absolute, base))) continue;
    try {
      if (fs.statSync(absolute).isFile()) return { path: absolute, error: null };
    } catch { /* 文件不存在则继续 */ }
  }
  return { path: null, error: "file is missing or outside approved download directories: " + downloadPath };
}

function validateUploadUrl(value) {
  let url;
  try { url = new URL(String(value || "")); } catch { throw new Error("uploadUrl is invalid"); }
  if (url.protocol !== "http:" || !LOOPBACK_HOSTS.has(url.hostname.toLowerCase()) || url.pathname !== CAPTURE_UPLOAD_PATH) {
    throw new Error("uploadUrl is not a loopback iBM capture endpoint");
  }
  if (url.username || url.password || url.hash) throw new Error("uploadUrl contains unauthorized components");
  const keys = [...url.searchParams.keys()];
  if (keys.length !== 1 || keys[0] !== "token" || !url.searchParams.get("token")) {
    throw new Error("uploadUrl must contain exactly one capture token");
  }
  return url;
}

function readResponseBody(response) {
  return new Promise((resolve, reject) => {
    const chunks = [];
    response.on("data", (chunk) => chunks.push(chunk));
    response.on("end", () => resolve(Buffer.concat(chunks).toString("utf8")));
    response.on("error", reject);
  });
}

async function upload(uploadUrl, filePath, fileName) {
  const url = validateUploadUrl(uploadUrl);
  const size = fs.statSync(filePath).size;
  if (size < 1 || size > MAX_FILE_BYTES) throw new Error("file size is outside the capture limit");
  const content = fs.readFileSync(filePath);
  const response = await new Promise((resolve, reject) => {
    const request = http.request({
      hostname: url.hostname,
      port: url.port || 80,
      path: url.pathname + url.search,
      method: "PUT",
      headers: {
        "Content-Type": "application/octet-stream",
        "X-File-Name": encodeURIComponent(fileName),
        "X-Capture-Source": "native-bridge",
        "Content-Length": content.length
      }
    }, resolve);
    request.setTimeout(300000, () => request.destroy(new Error("bridge upload timed out")));
    request.on("error", reject);
    request.end(content);
  });
  const body = await readResponseBody(response);
  try { return JSON.parse(body); }
  catch { return { ok: false, error: "server returned non-JSON response: " + body.slice(0, 200) }; }
}

async function main() {
  let message;
  try { message = readMessage(); }
  catch (error) {
    writeMessage({ ok: false, error: "malformed bridge message: " + error.message });
    return;
  }
  if (!message) {
    writeMessage({ ok: false, error: "empty message" });
    return;
  }
  if (message.cmd !== "upload") {
    writeMessage({ ok: false, error: "unsupported command: download bridge accepts upload only" });
    return;
  }

  const taskId = String(message.taskId || "");
  const uploadUrl = String(message.uploadUrl || "");
  const downloadPath = String(message.downloadPath || message.relativePath || "");
  const fileName = String(message.fileName || "") || path.basename(downloadPath.replace(/\\/g, "/"));
  if (!TASK_ID_RE.test(taskId)) {
    writeMessage({ ok: false, error: "invalid taskId" });
    return;
  }
  try { validateUploadUrl(uploadUrl); }
  catch (error) {
    writeMessage({ ok: false, taskId, error: error.message });
    return;
  }
  const resolved = resolveDownloadPath(downloadPath);
  if (resolved.error) {
    writeMessage({ ok: false, taskId, error: resolved.error });
    return;
  }
  try {
    const response = await upload(uploadUrl, resolved.path, fileName);
    const ok = Boolean(response.ok);
    writeMessage({ ok, taskId, response, error: ok ? null : String(response.error || "upload rejected") });
  } catch (error) {
    writeMessage({ ok: false, taskId, error: String(error.message || error) });
  }
}

module.exports = {
  BRIDGE_NAME,
  MAX_FILE_BYTES,
  CAPTURE_UPLOAD_PATH,
  downloadsDirs,
  resolveDownloadPath,
  insideDirectory,
  expandEnv,
  validateUploadUrl,
  readMessage,
  writeMessage,
  upload,
  main
};

if (require.main === module) {
  main().catch((error) => {
    log(String(error.message || error));
    writeMessage({ ok: false, error: String(error.message || error) });
  });
}
